torch/nested/_internal/triton_kernels/grouped_mm.py

```diff
@@ -0,0 +1,276 @@
+
+        offs_am = tile_m_idx * BLOCK_SIZE_M + tl.arange(0, BLOCK_SIZE_M)
+        offs_bn = tile_n_idx * BLOCK_SIZE_N + tl.arange(0, BLOCK_SIZE_N)
+
+        offs_am = offs_am % gm
+        offs_bn = offs_bn % gn
+
+        # Rematerialize on each loop
+        offs_k = tl.arange(0, BLOCK_SIZE_K)
+        a_ptrs = a_ptr + a_offset_0 * lda + (offs_am[:, None]) * lda + offs_k[None, :]
+        b_ptrs = b_ptr + batch_id * gk * gn + offs_k[:, None] * ldb + (offs_bn[None, :])
+        accumulator = tl.zeros((BLOCK_SIZE_M, BLOCK_SIZE_N), dtype=tl.float32)
+        for kk in range(0, tl.cdiv(gk, BLOCK_SIZE_K)):
+            # # # hint to Triton compiler to do proper loop pipelining
+            tl.multiple_of(a_ptrs, [16, 16])
+            tl.multiple_of(b_ptrs, [16, 16])
+            a = tl.load(
+                a_ptrs, mask=offs_k[None, :] < gk - kk * BLOCK_SIZE_K, other=0.0
+            )
+            b = tl.load(
+                b_ptrs, mask=offs_k[:, None] < gk - kk * BLOCK_SIZE_K, other=0.0
+            )
+            accumulator += tl.dot(a, b)  # , accumulator) #, "ieee")
+            a_ptrs += BLOCK_SIZE_K
+            b_ptrs += BLOCK_SIZE_K * ldb
+        c = accumulator.to(tl.float16)
+
+        # offs_cm = tile_m_idx * BLOCK_SIZE_M + tl.arange(0, BLOCK_SIZE_M)
+        # offs_cn = tile_n_idx * BLOCK_SIZE_N + tl.arange(0, BLOCK_SIZE_N)
+
+        # Trying to save registers by recomputing indices
+        offs_am = tile_m_idx * BLOCK_SIZE_M + tl.arange(0, BLOCK_SIZE_M)
+        offs_bn = tile_n_idx * BLOCK_SIZE_N + tl.arange(0, BLOCK_SIZE_N)
+
+        a_offset_0 = tl.load(a_offsets_ptr + batch_id, eviction_policy="evict_last").to(
+            tl.int32
+        )
+        a_offset_1 = tl.load(
+            a_offsets_ptr + batch_id + 1, eviction_policy="evict_last"
+        ).to(tl.int32)
+        gm = a_offset_1 - a_offset_0
+
+        c_ptrs = c_ptr + a_offset_0 * ldc + ldc * offs_am[:, None] + offs_bn[None, :]
+        c_mask = (offs_am[:, None] < gm) & (offs_bn[None, :] < gn)
+
+        # assumes full tile for now
+        tl.store(c_ptrs, c, mask=c_mask)
+
+
+def group_gemm_fn_kernel(a_values, a_offsets, max_M, tensor_b, c_values, config):
+    B, K, N = tensor_b.shape
+    group_size = a_offsets.size(0)
+    grid = (
+        triton.cdiv(max_M, config["BLOCK_SIZE_M"])
+        * triton.cdiv(N, config["BLOCK_SIZE_N"]),
+        group_size,
+    )
+    grouped_matmul_kernel[grid](
+        tensor_b,
+        group_size,
+        N,
+        K,
+        tensor_b.stride(1),
+        a_offsets,
+        a_values,
+        a_values.stride(0),
+        c_values.stride(0),
+        c_values,
+        max_M,
+        BLOCK_SIZE_M=config["BLOCK_SIZE_M"],
+        BLOCK_SIZE_N=config["BLOCK_SIZE_N"],
+        BLOCK_SIZE_K=config["BLOCK_SIZE_K"],
+        GROUP_SIZE_M=config["GROUP_SIZE_M"],
+        num_stages=config["num_stages"],
+        num_warps=config["num_warps"],
+    )
+
+    return c_values
+
+
+BEST_CONFIGS = {}
+BEST_CONFIGS[
```

Comment:
Should we make these gpu-dependent?

Reply:
Right, so this is tricky. What if the GPU has different settings (e.g. power consumption) than expected? I was thinking in an upcoming iteration to allow users to clear or pass these configs themselves and we just provide a baseline.